File: sdk/python/tennetctl/_transport.py
from __future__ import annotations

import asyncio
from typing import Any

import httpx

from .errors import NetworkError, map_error

SESSION_COOKIE = "tnt_session"

RETRY_STATUSES = {502, 503, 504}
RETRY_BACKOFFS = (0.5, 1.0, 2.0)  # seconds; len = max retries after first attempt
MAX_ATTEMPTS = len(RETRY_BACKOFFS) + 1


class Transport:
    """Async HTTP transport: bearer auth, envelope parse, typed errors, retry."""

    def __init__(
        self,
        base_url: str,
        *,
        api_key: str | None = None,
        session_token: str | None = None,
        timeout: float = 30.0,
        client: httpx.AsyncClient | None = None,
    ) -> None:
        self._base_url = base_url.rstrip("/")
        self._api_key = api_key
        self._session_token = session_token
        cookies: dict[str, str] | None = None
        if session_token:
            cookies = {SESSION_COOKIE: session_token}
        self._client = client or httpx.AsyncClient(
            base_url=self._base_url,
            timeout=timeout,
            cookies=cookies,
        )
        self._owns_client = client is None

    @property
    def session_token(self) -> str | None:
        return self._session_token

    def set_session_token(self, token: str | None) -> None:
        self._session_token = token
        if token:
            self._client.cookies.set(SESSION_COOKIE, token)
        else:
            try:
                self._client.cookies.delete(SESSION_COOKIE)
            except KeyError:
                pass

    async def request(
        self,
        method: str,
        path: str,
        *,
        json: Any = None,
        params: dict[str, Any] | None = None,
        headers: dict[str, str] | None = None,
    ) -> Any:
        hdrs: dict[str, str] = dict(headers or {})
        if self._api_key:
            hdrs["Authorization"] = f"Bearer {self._api_key}"

        last_exc: Exception | None = None
        for attempt in range(MAX_ATTEMPTS):
            try:
                response = await self._client.request(
                    method.upper(),
                    path,
                    json=json,
                    params=params,
                    headers=hdrs,
                )
            except (httpx.ConnectError, httpx.ReadTimeout, httpx.ConnectTimeout, httpx.RemoteProtocolError) as exc:
                last_exc = exc
                if attempt < MAX_ATTEMPTS - 1:
                    await asyncio.sleep(RETRY_BACKOFFS[attempt])
                    continue
                raise NetworkError(str(exc), code="NETWORK", status=None) from exc

            if response.status_code in RETRY_STATUSES and attempt < MAX_ATTEMPTS - 1:
                await asyncio.sleep(RETRY_BACKOFFS[attempt])
                continue

            return self._parse_response(response)

        # should not reach here — loop either returns or raises
        raise NetworkError("retry loop exhausted", code="NETWORK") from last_exc

    def _parse_response(self, response: httpx.Response) -> Any:
        # 204 No Content — no body to parse
        if response.status_code == 204:
            return None

        try:
            envelope = response.json()
        except ValueError:
            envelope = None

        if 200 <= response.status_code < 300:
            if isinstance(envelope, dict) and envelope.get("ok") is True:
                return envelope.get("data")
            # Non-envelope 2xx (e.g., raw JSON from upstream) — return body as-is
            return envelope

        # Error path
        if not isinstance(envelope, dict):
            envelope = {"ok": False, "error": {"code": "HTTP_ERROR", "message": response.text or f"HTTP {response.status_code}"}}
        raise map_error(response.status_code, envelope)

    async def close(self) -> None:
        if self._owns_client:
            await self._client.aclose()

    async def __aenter__(self) -> Transport:
        return self

    async def __aexit__(self, *exc_info: Any) -> None:
        await self.close()

File: sdk/typescript/src/transport.ts
import { NetworkError, mapError } from "./errors.js";

export const SESSION_COOKIE = "tnt_session";

const RETRY_STATUSES = new Set([502, 503, 504]);
const RETRY_BACKOFFS_MS = [500, 1000, 2000] as const;
const MAX_ATTEMPTS = RETRY_BACKOFFS_MS.length + 1;

export type FetchFn = typeof fetch;

export interface TransportOptions {
  baseUrl: string;
  apiKey?: string;
  sessionToken?: string;
  timeoutMs?: number;
  fetchFn?: FetchFn;
  sleepFn?: (ms: number) => Promise<void>;
}

const defaultSleep = (ms: number): Promise<void> =>
  new Promise((resolve) => setTimeout(resolve, ms));

export class Transport {
  private readonly baseUrl: string;
  private readonly apiKey?: string;
  private sessionToken?: string;
  private readonly timeoutMs: number;
  private readonly fetchFn: FetchFn;
  private readonly sleepFn: (ms: number) => Promise<void>;

  constructor(opts: TransportOptions) {
    this.baseUrl = opts.baseUrl.replace(/\/+$/, "");
    this.apiKey = opts.apiKey;
    this.sessionToken = opts.sessionToken;
    this.timeoutMs = opts.timeoutMs ?? 30_000;
    this.fetchFn = opts.fetchFn ?? fetch;
    this.sleepFn = opts.sleepFn ?? defaultSleep;
  }

  getSessionToken(): string | undefined {
    return this.sessionToken;
  }

  setSessionToken(token: string | undefined): void {
    this.sessionToken = token;
  }

  async request<T = unknown>(
    method: string,
    path: string,
    opts: {
      body?: unknown;
      params?: Record<string, string | number | boolean | undefined>;
      headers?: Record<string, string>;
    } = {},
  ): Promise<T> {
    const url = this.buildUrl(path, opts.params);
    const headers = this.buildHeaders(opts.body !== undefined, opts.headers);

    for (let attempt = 0; attempt < MAX_ATTEMPTS; attempt++) {
      const controller = new AbortController();
      const timer = setTimeout(() => controller.abort(), this.timeoutMs);

      let response: Response;
      try {
        response = await this.fetchFn(url, {
          method: method.toUpperCase(),
          headers,
          body: opts.body !== undefined ? JSON.stringify(opts.body) : undefined,
          credentials: "include",
          signal: controller.signal,
        });
      } catch (err) {
        clearTimeout(timer);
        if (attempt < MAX_ATTEMPTS - 1) {
          await this.sleepFn(RETRY_BACKOFFS_MS[attempt]!);
          continue;
        }
        throw new NetworkError(err instanceof Error ? err.message : String(err), {
          code: "NETWORK",
          status: null,
        });
      } finally {
        clearTimeout(timer);
      }

      if (RETRY_STATUSES.has(response.status) && attempt < MAX_ATTEMPTS - 1) {
        await this.sleepFn(RETRY_BACKOFFS_MS[attempt]!);
        continue;
      }

      return this.parseResponse<T>(response);
    }

    throw new NetworkError("retry loop exhausted", { code: "NETWORK" });
  }

  private buildUrl(
    path: string,
    params?: Record<string, string | number | boolean | undefined>,
  ): string {
    const url = new URL(path.startsWith("/") ? path.slice(1) : path, this.baseUrl + "/");
    if (params) {
      for (const [k, v] of Object.entries(params)) {
        if (v !== undefined) url.searchParams.set(k, String(v));
      }
    }
    return url.toString();
  }

  private buildHeaders(
    hasBody: boolean,
    extra?: Record<string, string>,
  ): Record<string, string> {
    const headers: Record<string, string> = { Accept: "application/json" };
    if (hasBody) headers["Content-Type"] = "application/json";
    if (this.apiKey) headers["Authorization"] = `Bearer ${this.apiKey}`;
    if (extra) {
      for (const [k, v] of Object.entries(extra)) {
        headers[k] = v;
      }
    }
    // Cookie handled by fetch with credentials: "include" in browser;
    // in Node runtime the server must be same-origin or configure CORS.
    return headers;
  }

  private async parseResponse<T>(response: Response): Promise<T> {
    if (response.status === 204) {
      return undefined as T;
    }

    let envelope: unknown = null;
    const text = await response.text();
    if (text.length > 0) {
      try {
        envelope = JSON.parse(text);
      } catch {
        envelope = null;
      }
    }

    if (response.status >= 200 && response.status < 300) {
      if (
        envelope &&
        typeof envelope === "object" &&
        (envelope as { ok?: unknown }).ok === true
      ) {
        return (envelope as { data?: unknown }).data as T;
      }
      return envelope as T;
    }

    const body =
      envelope ??
      {
        ok: false,
        error: { code: "HTTP_ERROR", message: text || `HTTP ${response.status}` },
      };
    throw mapError(response.status, body);
  }
}
